packages/cdk-assets/lib/handlers/files.ts

```diff
@@ -0,0 +1,77 @@
+import { FileAssetPackaging } from '@aws-cdk/asset-manifest-schema';
+import * as fs from 'fs-extra';
+import * as path from 'path';
+import { FileManifestEntry } from '../asset-manifest';
+import { IAws } from "../aws-operations";
+import { zipDirectory } from '../private/archive';
+import { IAssetHandler, MessageSink } from "../private/asset-handler";
+import { replaceAwsPlaceholders } from "../private/placeholders";
+
+export class FileAssetHandler implements IAssetHandler {
+  private readonly fileCacheRoot: string;
+
+  constructor(
+    private readonly root: string,
+    private readonly asset: FileManifestEntry,
+    private readonly aws: IAws,
+    private readonly message: MessageSink) {
+    this.fileCacheRoot = path.join(root, '.cache');
+  }
+
+  public async publish(): Promise<void> {
+    const destination = await replaceAwsPlaceholders(this.asset.destination, this.aws);
+    const s3Url = `s3://${destination.bucketName}/${destination.objectKey}`;
+
+    const s3 = this.aws.s3Client(destination);
+    this.message(`Check ${s3Url}`);
+    if (await objectExists(s3, destination.bucketName, destination.objectKey)) {
+      this.message(`Found ${s3Url}`);
+      return;
+    }
+
+    const publishFile = await this.package();
+    const contentType = this.asset.source.packaging === FileAssetPackaging.ZIP_DIRECTORY ? 'application/zip' : undefined;
+
+    this.message(`Upload ${s3Url}`);
+    await s3.putObject({
```

Comment:
May I suggest to use [`s3.upload`](https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/S3.html#upload-property) here? It has the following advantages over `s3.putObject`:
* Intelligent concurrent handling of parts if the payload is large enough
* It is the only operation for which the SDK can retry requests with stream bodies

The function signature is the same (except for the optional `options`) and it requires the same IAM permissions.